Update rinja to 0.3.2 and do some template cleanup

GuillaumeGomez opened the pull request:
The `is (not) defined` feature was improved, allowing to start some cleanups in the templates. I'm not completely done yet but I think it's big enough for now.

syphar commented:
This is an awesome improvement, thanks!

GuillaumeGomez commented on src/web/rustdoc.rs:
```diff
@@ -257,22 +258,6 @@ pub(crate) async fn rustdoc_redirector_handler(
 }
 
-#[derive(Debug, Clone)]
-pub struct RustdocPage {
```
Because of the `filters` module. I can import it here though. As you prefer.

GuillaumeGomez replied:
Moved it back.

GuillaumeGomez commented on src/web/crate_details.rs:
```diff
@@ -32,6 +32,8 @@ use std::sync::Arc;
 // TODO: Add target name and versions
 
+#[derive(Template)]
+#[template(path = "crate/details.html")]
```
Is there a reason why in particular? `CrateDetailsPage` is exactly the same type and use most of the fields from `CrateDetails`, so not sure having this separation makes much sense. If you look at each commit, I started by moving one field after another directly into `CrateDetails` but at some point, I just merged both types because it didn't make sense.

syphar replied:
having thought about it, it's still feels consusing for users of that struct that it carries a `csp_nonce` around, which is only used in one view.

Other (current) users of that struct (`target_redirect_handler`, `rustdoc_html_server_handler`, `rustdoc_redirector_handler`) would also get it and it would be empty?

Kijewski commented on templates/crate/details.html:
```diff
@@ -131,61 +128,61 @@
+                {%- if is_library == Some(false) -%}
                     <div class="warning">
-                        {{ details.name }}-{{ details.version }} is not a library.
+                        {{ self.name }}-{{ version }} is not a library.
```
With rinja v0.3.2 you now can `%s/self.name/name/g`, can't you?

GuillaumeGomez replied:
Yes I can!

GuillaumeGomez commented:
I moved back `RustdocPage` to where it originally was and cleaned up the code thanks to the fix released in rinja 0.3.2.

Ready for another review round. :)

Kijewski commented on src/web/page/templates.rs:
```diff
@@ -1,45 +1,31 @@
+#[derive(Template)]
+#[template(path = "rustdoc/head.html")]
+pub struct Head<'a> {
+    rustdoc_css_file: &'a Option<String>,
```
I would rather use `Option<&'a str>`, to omit the second indirection. (https://rust-lang.github.io/rust-clippy/stable/index.html#/ref_option_ref)

Kijewski commented on templates/rustdoc/topbar.html:
```diff
@@ -34,7 +39,7 @@
                     {%- if metadata.req_version|to_string == "latest" -%}
                     <li class="pure-menu-item">
-                        <a href="{{permalink_path()|safe}}" class="pure-menu-link description" id="permalink" title="Get a link to this specific version">
+                        <a href="{% if permalink_path is defined %}{{permalink_path|safe}}{% endif %}" class="pure-menu-link description" id="permalink" title="Get a link to this specific version">
```
Shouldn't the if-block be outside if the `href` attribute? If `permalink_path` is undefined, then an empty `href=""` is rendered. I actually don't know what its semantics are.

GuillaumeGomez replied:
It's the goal actually. :wink:

Kijewski commented on templates/rustdoc/topbar.html:
```diff
@@ -198,7 +203,7 @@
 
         <li class="pure-menu-item">
-            <a href="{{ latest_path|safe }}" class="pure-menu-link warn"
+            <a href="{% if latest_path is defined %}{{ latest_path|safe }}{% endif %}" class="pure-menu-link warn"
```
Are comment as above: `{%if …%}href="…" {%endif%}`

GuillaumeGomez replied:
Same answer. :wink:

Kijewski reviewed (changes requested):
I hope reviews from non-members are welcome. If not, please simply ignore me. :)

GuillaumeGomez commented:
> I hope reviews from non-members are welcome. If not, please simply ignore me. :)

It's very welcome. :)

syphar reviewed (commented):
sorry for the delay, chaotic times. 

I'm really happy about these improvements and the progress that `rinja` makes. 

I also still feel like that `CrateDetails` and `CrateDetailsPage` should stay separate.

GuillaumeGomez commented:
Fair enough, I'll split them again.

GuillaumeGomez commented:
Split back `CrateDetails` struct. :)
